src/engine/strat_engine/range_alloc.rs

```diff
@@ -0,0 +1,232 @@
+                (Some((off, len)), false) => {
+                    used.push((Sectors(off), Sectors(len)));
+                    None
+                }
+            }
+        });
+
+        used
+    }
+
+    /// Get a list of (offset, length) segments that are not in use
+    fn avail_ranges(&self) -> Vec<(Sectors, Sectors)> {
+        let mut free = Vec::new();
+
+        // Insert an entry to mark the end so the fold works correctly
+        let mut used = self.used_ranges();
+        used.push((self.limit, Sectors(0)));
+
+        used.into_iter()
+            .fold(Sectors(0), |prev_end, (start, len)| {
+                if prev_end < start {
+                    free.push((prev_end, start - prev_end))
+                }
+                start + len
+            });
+
+        free
+    }
+
+    /// Attempt to allocate. Returns number of sectors allocated (may
+    /// be less than request, including zero) and a Vec<(offset,
+    /// length)> of sectors successfully allocated.
+    pub fn request(&mut self, amount: Sectors) -> (Sectors, Vec<(Sectors, Sectors)>) {
+        let mut segs = Vec::new();
+        let mut needed = amount;
+
+        for (start, len) in self.avail_ranges() {
+            if needed == Sectors(0) {
+                break;
+            }
+
+            let to_use = min(needed, len);
+
+            let used_range = (start, to_use);
+            segs.push(used_range);
+            self.insert_ranges(&[used_range]);
+
+            needed = needed - to_use;
+        }
+
+        (amount - needed, segs)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_allocator_allocations() {
+        let mut allocator = RangeAllocator::new(Sectors(128));
+
+        assert_eq!(allocator.used(), Sectors(0));
+        assert_eq!(allocator.available(), Sectors(128));
+
+        allocator.insert_ranges(&[(Sectors(10), Sectors(100))]);
+
+        assert_eq!(allocator.used(), Sectors(100));
+        assert_eq!(allocator.available(), Sectors(28));
+
+        let request = allocator.request(Sectors(50));
+        assert_eq!(request.0, Sectors(28));
+        assert_eq!(allocator.used(), Sectors(128));
+        assert_eq!(allocator.available(), Sectors(0));
+        assert_eq!(request.1.len(), 2);
+
+        let good_remove_ranges = [(Sectors(21), Sectors(20)), (Sectors(41), Sectors(40))];
+        allocator.remove_ranges(&good_remove_ranges);
+        assert_eq!(allocator.used(), Sectors(68));
+        assert_eq!(allocator.available(), Sectors(60));
+    }
+
+    #[test]
+    #[should_panic]
+    fn test_allocator_failures() {
+        let mut allocator = RangeAllocator::new(Sectors(128));
+
+        let _request = allocator.request(Sectors(128));
+
+        // overlap
+        let bad_remove_ranges = [(Sectors(21), Sectors(20)), (Sectors(40), Sectors(40))];
+        allocator.remove_ranges(&bad_remove_ranges);
+    }
+
+    #[test]
+    #[should_panic]
```

Comment:
Arrgh, I updated this from the version you commented on, and am now lost trying to tie your comments to the appropriate test case. Adding these comments is 👍. I think it might be quicker for you to push a commit to this PR that makes the additions, would you be so kind, please? 😃